flax/linen/normalization.py

```diff
@@ -46,25 +53,27 @@ def _compute_stats(x: Array, axes: Axes,
   This implementation takes care of a few important details:
   - Computes in float32 precision for half precision inputs
   -  mean and variance is computable in a single XLA fusion,
-    by using Var = E[x^2] - E[x]^2 instead of Var = E[(x - E[x])^2]).
+    by using Var = E[|x|^2] - |E[x]|^2 instead of Var = E[|x - E[x]|^2]).
   - Clips negative variances to zero which can happen due to
     roundoff errors. This avoids downstream NaNs.
   - Supports averaging across a parallel axis and subgroups of a parallel axis
     with a single `lax.pmean` call to avoid latency.
   """
-  x = jnp.asarray(x, jnp.float32)
+  # promote x to at least float32, this avoids half precision computation
+  # but preserves double or complex floating points
+  x = jnp.asarray(x, jnp.promote_types(jnp.float32, jnp.result_type(x)))
```

Comment:
@marcvanzee  this is the part that affects both complex and float64. We cast x to float32 before computing the statistics. I changed that to promoting to at least float32 so we don't downcast f64 and complex numbers